src/main/cpp/src/datetime_utils.cuh

```diff
@@ -23,91 +23,216 @@
+    return (month == 4 || month == 6 || month == 9 || month == 11) ? 30 : 31;
+  }
 
   /**
-   * Get epoch day. Can handle years in the range [-1,000,000, 1,000,000].
-   * Spark supports years range: [-290307, 294247], so this is enough.
-   * Refer to cuda::std::chrono::year_month_day, its range is [-32,767 , 32,767],
-   * because of year is short type in cuda::std::chrono, so we can not use it.
+   * @brief Calculate the number of days since epoch 1970-01-01 from date.
+   * Refer to https://howardhinnant.github.io/date_algorithms.html#days_from_civil.
+   * Unlike cuda::std::chrono::year_month_day only supports year range [-32,767 , 32,767],
+   * this implementation supports all int years.
    */
-  __device__ int32_t to_epoch_day() const
+  __device__ static int64_t to_epoch_day(int const year, int const month, int const day)
   {
     int32_t y          = year - (month <= 2);
     const int32_t era  = (y >= 0 ? y : y - 399) / 400;
     const uint32_t yoe = static_cast<uint32_t>(y - era * 400);                           // [0, 399]
     const uint32_t doy = (153 * (month > 2 ? month - 3 : month + 9) + 2) / 5 + day - 1;  // [0, 365]
     const uint32_t doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;  // [0, 146096]
-    return era * 146097 + doe - 719468;
+    return era * 146097L + doe - 719468L;
   }
 
-  // 4-6 digits
-  int32_t year;
-
-  // 1-12
-  int32_t month;
+  /**
+   * @brief Calculate year/month/day from the number of days since epoch 1970-01-01.
+   * Refer to https://howardhinnant.github.io/date_algorithms.html#civil_from_days.
+   * This implementation supports all int values as `epoch_day`.
+   * @param epoch_day the number of days since 1970-01-01
+   * @param year[out]      year output year
+   * @param month[out]     month output month
+   * @param day[out]       day output day
+   */
+  __device__ static void to_date(int32_t const epoch_day, int& year, int& month, int& day)
+  {
+    int64_t z = static_cast<int64_t>(epoch_day);
+    z += 719468;
+    const int32_t era  = static_cast<int32_t>((z >= 0 ? z : z - 146096) / 146097);
+    const uint32_t doe = static_cast<uint32_t>(z - era * 146097);                // [0, 146096]
+    const uint32_t yoe = (doe - doe / 1460 + doe / 36524 - doe / 146096) / 365;  // [0, 399]
+    const int32_t y    = static_cast<uint32_t>(yoe) + era * 400;
+    const uint32_t doy = doe - (365 * yoe + yoe / 4 - yoe / 100);                // [0, 365]
+    const uint32_t mp  = (5 * doy + 2) / 153;                                    // [0, 11]
+    const uint32_t d   = doy - (153 * mp + 2) / 5 + 1;                           // [1, 31]
+    const uint32_t m   = mp < 10 ? mp + 3 : mp - 9;                              // [1, 12]
+    year               = y + (m <= 2);
+    month              = static_cast<uint8_t>(m);
+    day                = static_cast<uint8_t>(d);
+  }
 
-  // 1-31; it is 29 for leap February, or 28 for regular February
-  int32_t day;
+  /**
+   * @brief Is the month/day is valid.
```

Comment:
How about "If the month/day is valid." of "Is the Spark date valid."
This also applies to to `is_valid_date_for_date`, `is_valid_date_for_timestamp`, and `is_valid_time `